### PR TITLE
Removed Tagged Classes

### PR DESCRIPTION
**Pull request by InsertCreativityHere**
This PR disallows the tagging of any type that uses classes at the compiler level (it reports an error when you do so) and also removes all our usage and testing of tagged classes from the repository.

### CODE REVIEW

File: cpp/src/Slice/Grammar.y [hunk 1]
```diff
@@ -2089,6 +2098,10 @@ tagged_type
         type = new Optional(TypePtr::dynamicCast($2));
         unit->error("only optional types can be tagged");
     }
+    if (type->usesClasses())
```
Comment (pepone) [inline cpp/src/Slice/Grammar.y:2101]: Maybe we should have separate error messages for class types, and types that use classes. So rather than output

```
TaggedMembers.ice:112: type '::Ice::AnyClass' cannot be tagged, as it uses classes
```

We can have

```
TaggedMembers.ice:112: type '::Ice::AnyClass' class cannot be tagged
```

File: matlab/test/Ice/optional/ClientPrivate.ice [hunk 2]
```diff
@@ -12,12 +12,12 @@ module Test
...
-    string ds;
-    tag(990) StringSeq? seq;
-    tag(1000) A? ao;
-}
+// class D : B
+// {
+//     string ds;
+//     tag(990) StringSeq? seq;
+//     tag(1000) A? ao;
+// }
```
Comment (pepone) [inline matlab/test/Ice/optional/ClientPrivate.ice:20]: Shouldn't we remove this?

**Review by pepone — APPROVED** [review]
Looks good, just a minor suggestion regarding the error messages

File: cpp/src/Slice/Grammar.y (hunk 1 above)
Comment (InsertCreativityHere) [reply cpp/src/Slice/Grammar.y:2101]: Yeah, having better error messages is definitely a good thing!
But I'll probably change it to:
`class '::Test::Foo' cannot be tagged`
if the type being tagged is directly a class, and then keep the current error message for if it uses classes indirectly.

File: matlab/test/Ice/optional/ClientPrivate.ice (hunk 2 above)
Comment (InsertCreativityHere) [reply matlab/test/Ice/optional/ClientPrivate.ice:20]: Maybe, but we also have the code commented out in a couple other language mappings already, so I just followed their lead. I'm not sure if we ever planned on un-commenting it, but if it's definitely dead code, I'll remove it!

File: cpp/src/Slice/Grammar.y (hunk 1 above)
Comment (bernardnormier) [reply cpp/src/Slice/Grammar.y:2101]: I would also rework the "uses classes" term - it's an internal API, not a terminology meaningful to the user. Something like:
```
cannot use `AnyClass' as a tagged data member
cannot use class `FooBar' as a tagged parameter
cannot use `FooBar` as a tagged parameter because it holds directly or indirectly a class data member
```

**Review by bernardnormier — APPROVED** [review]
Looks good, but should improve error message.

File: cpp/src/Slice/Grammar.y (hunk 1 above)
Comment (InsertCreativityHere) [reply cpp/src/Slice/Grammar.y]: I can drop the 'class' in front for when it's type 'AnyClass' instead of a concrete class. I agree that is more correct.

But I think I prefer my `class 'Foo' cannot be tagged` to your first 2 suggested alternatives. It's more clear to the user that the issue is 'classes cannot be tagged', instead of making it seem like something specific to either parameters or data members. And it isn't even that 'using' it is the issue, just the act of 'tagging' it is erroneous.

I've tried thinking for alternatives to my second error message, but the best I can come up with is:
`'Foo' cannot be tagged, as it contains classes`
I think 'contains' is definitely better than 'uses'... But all the other ones I've thought up, and the one you proposed, seem way too wordy to me.

File: matlab/test/Ice/optional/ClientPrivate.ice (hunk 2 above)
Comment (bernardnormier) [reply matlab/test/Ice/optional/ClientPrivate.ice:20]: Please remove it.

Comment (InsertCreativityHere) [reply matlab/test/Ice/optional/ClientPrivate.ice:20]: Removed!

File: cpp/src/Slice/Grammar.y (hunk 1 above)
Comment (InsertCreativityHere) [reply cpp/src/Slice/Grammar.y]: Good point, I've spent too long in the compiler where we tie the tag to the type, and not the member...
That being said, how about:
`member 'myparam' cannot be tagged, since it's type 'AnyClass' is a class`
`member 'myparam' cannot be tagged, since it's type 'Foo' is a class`
`member 'myparam' cannot be tagged, since it's type 'Foo' contains a class`
It makes it clear that it's the member that's being tagged, but also makes it clear that the underlying reason is that it's type contains classes one way or another.

Comment (bernardnormier) [reply cpp/src/Slice/Grammar.y]: Can you output "data member", "parameter", "return type member" as appropriate instead of always "member"?

Comment (InsertCreativityHere) [reply cpp/src/Slice/Grammar.y]: Yeah, it's definitely possible, it'll just take some extra code to check what the enclosing container is.
